Userguide overhaul (#316)
cylc/cylc-doc: src/user-guide/writing-workflows/runtime.rst

> .. TODO - consider a re-write once dynamic platform selection done

Reviewer: https://github.com/cylc/cylc-doc/issues/336

> .. TODO - consider a re-write once dynamic platform selection done

Reviewer: https://github.com/cylc/cylc-doc/issues/337

Job hosts are always selected dynamically, for the chosen platform.

Remote Task Job Log Directories
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Task job stdout and stderr streams are written to :term:`log files <job log>`
under the workflow :term:`run directory` (see :ref:`WhitherStdoutAndStderr`).
For remote tasks the same directory is used, on the job host.

.. _ImplicitTasks:

Implicit Tasks
--------------

An implicit task is one that appears in the graph but is not defined under
:cylc:conf:`flow.cylc[runtime]`.

Depending on the value of :cylc:conf:`flow.cylc[scheduler]allow implicit tasks`
Cylc can automatically create default task definitions for these, to submit
local dummy jobs that just return the standard job status messages.

Implicit tasks can be used to mock up functional workflows very quickly. A
default ``script`` can be added to the root family, e.g. to slow job execution
down a little. Here is a complete workflow definition using implicit tasks:

.. code-block:: cylc

   [scheduler]
       allow implicit tasks = True
   [scheduling]
       [[graph]]
           R1 = "prep => run-a & run-b => done"
   [runtime]
       [[root]]
           script = "sleep 10"

.. warning::

>   Implicit tasks are somewhat dangerous because they can easily be created by
>   mistake: misspelling a task's name divorces it from its ``runtime`` definition.

Reviewer 2: ```suggestion
   Implicit tasks are somewhat dangerous because they can easily be created by
   mistake; misspelling a task's name divorces it from its ``runtime`` definition.
```

Author: Applied the indent, but I think the colon is correct here. The second clause explains how implicit tasks can be created by mistake.

       [[graph]]
           R1 = """
               bad => cheese
               flaky => whizz
            """
   [runtime]
       [[bad]]
           # retry 3 times then fail
           script = """
               sleep 10
               false
           """
           execution retry delays = 3*PT10S
       [[flaky]]
           # retry 3 times then succeed
           script = """
               sleep 10
               test $CYLC_TASK_TRY_NUMBER -gt 3
           """
           execution retry delays = 3*PT10S
       [[cheese, whizz]]
           script = "sleep 10"

Aborting a Retry Sequence
^^^^^^^^^^^^^^^^^^^^^^^^^

To prevent a waiting task from retrying, remove it from the scheduler.
If the example above is installed as ``demo``:

.. code-block:: console

   $ cylc remove demo flaky.1

If a task with retries gets *killed* while running, it goes to the ``held``
state so you decide whether to release it and continue the retry
sequence, or abort.

.. code-block:: console

   $ cylc kill demo flaky.1  # flaky.1 goes to held state post kill
   $ cylc release demo flaky.1  # release to continue retrying
   $ cylc remove demo flaky.1  # or remove the task to abort retries

> If you want ``whizz`` to triggr downstream despite ``flaky.1`` being removed

Reviewer 2: ```suggestion
If you want ``whizz`` to trigger downstream tasks despite ``flaky.1`` being removed
```

see :ref:`Managing External Command Execution`.

.. cylc-scope::

Event-Specific Handlers
^^^^^^^^^^^^^^^^^^^^^^^

Event-specific handlers can be configured in ``<event> handlers``
under :cylc:conf:`[runtime][<namespace>][events]`, where ``<event>`` can be:

  - **submitted**

    - task job submitted
  - **started**

    - task job started running
  - **succeeded**

    - task job succeeded
  - **failed**

    - task job failed
  - **retry**

    - task job just failed but will retry after a configured delay
  - **submission failed**

    - task job submission failed
  - **submission retry**

    - task job submission failed but will retry after a configured delay
  - **submission timeout**

    - task job timed out in the submitted state
  - **execution timeout**

    - task job timed out in the running state
  - **warning**

    - scheduler received a WARNING message from the task job
  - **critical**

    - scheduler received a CRITICAL message from the task job
  - **expired**

    - task expired (too far behind the clock to bother submitting)
  - **late**

    - task will run later than expected (by some configured interval)
  - **custom**

>     - scheduler received a custom event message received from the task job

Reviewer 2: ```suggestion
    - scheduler received a custom event message from the task job
```

  - **<name>**

>     - scheduler recieved the <name> output received from the task job

Reviewer 2: ```suggestion
    - scheduler received the <name> output from the task job
```

The value should be a list of commands, command lines, or command line
templates (see below) to call if the specified event is triggered.

General Event Handlers
^^^^^^^^^^^^^^^^^^^^^^

Alternatively you can configure ``handler events`` and ``handlers``

> under :cylc:conf:`[runtime][<namespace>][events]`, where the former a list of
> events (as above), and the latter a list of commands, command lines,
> lines or command line templates (see below) to call if any of the specified

Reviewer 2: ```suggestion
under :cylc:conf:`[runtime][<namespace>][events]`, where the former is a list of
events (as above), and the latter is a list of commands, command lines,
or command line templates (see below) to call if any of the specified
```

Author: Took the first "is"!

events are triggered.

.. cylc-scope::

Event Handler Command Line Arguments
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Cylc makes event data available to event handlers by command line argument
templating: workflow name; task ID, name, cycle point, message, submit
number; and any :cylc:conf:`workflow <[meta]>` or

> :cylc:conf:`task <[runtime][<namespace>][meta]>` item.

Reviewer 2: Not sure this sentence makes sense

Author: Fixed.

  currently implemented.

If a real time (clock-triggered) workflow performs fairly consistently from one
cycle to the next, you may want to be notified when certain tasks are running
late with respect the time they normally trigger in each cycle.

Cylc can generate a *late* event if a task has not triggered by a given offset
from its cycle point in real time. For example, if a task ``forecast`` normally
triggers at 30 minutes after cycle point, a late event could be configured like this:

.. code-block:: cylc

   [runtime]
      [[forecast]]
           script = run-model.sh
           [[[events]]]
               late offset = PT40M  # allow a 10 minute delay
               late handler = my-handler %(message)s

.. warning::
   Late offset intervals are not computed automatically so be careful to update
   them after any workflow change that affects triggering times.

Workflow Events
^^^^^^^^^^^^^^^

Cylc also supports general and event-specific workflow event handling so
you can run an event handler or send and email when the scheduler starts up or
shuts down, for example.

Workflow events are configured in :cylc:conf:`[scheduler][events]`. The list of
events is:

>   - startup
> 
>     - the scheduler started running the workflow
>   - shutdown
> 
>     - the workflow finished and the scheduler will shut down
>   - abort
> 
>     - the scheduler will shut down even though the workflow didn't finish
>   - workflow timeout
> 
>     - the workflow run timed out
>   - stall
> 
>     - the workflow stalled
>   - stall timeout
> 
>     - the workflow timed out after stalling
>   - inactivity timeout
> 
>     - the workflow timed out with no activity

Reviewer: Blockquote:

```suggestion
- startup

  - the scheduler started running the workflow
- shutdown

  - the workflow finished and the scheduler will shut down
- abort

  - the scheduler will shut down even though the workflow didn't finish
- workflow timeout

  - the workflow run timed out
- stall

  - the workflow stalled
- stall timeout

  - the workflow timed out after stalling
- inactivity timeout

  - the workflow timed out with no activity
```

You can also tell the scheduler to abort the run on certain workflow events,
with the following settings:

>   - abort on stall timeout
>   - abort on inactivity timeout
>   - abort on workflow timeout

Reviewer: blockquote:

```suggestion
- abort on stall timeout
- abort on inactivity timeout
- abort on workflow timeout
```

E.g. to send an email on task failed, retry, and a custom message event:

.. code-block:: cylc

   [runtime]
       [[foo]]
           script = """
               test ${CYLC_TASK_TRY_NUMBER} -eq 3
               cylc message -- "${CYLC_WORKFLOW_ID}" "${CYLC_TASK_JOB}" 'oopsy daisy'
           """
           execution retry delays = PT0S, PT30S
           [[[events]]]
               mail events = failed, retry, oops
           [[[outputs]]]
               oops = oopsy daisy

By default, event emails will be sent to the current user with:

- ``to:`` set as ``$USER``
- ``from:`` set as ``notifications@$(hostname)``
- SMTP server at ``localhost:25``

These can be configured using the settings:

.. cylc-scope:: flow.cylc[runtime][<namespace>]

- :cylc:conf:`[mail]to` (list of email addresses)
- :cylc:conf:`[mail]from`

.. cylc-scope::

The scheduler batches events over a 5 minute interval, by default, to avoid
flooding your Inbox if many events occur in a short time. The batching interval
can be configured with :cylc:conf:`[scheduler][mail]task event batch interval`.

.. _Late Events:

Late Events
^^^^^^^^^^^

.. warning::

>   The scheduler can only check for lateness once a task has appeared in its
>   active task window. In Cylc 8 this is usually when the task is actually
>   ready to run, which severely limits the usefulness of late events as
>   currently implemented.

Reviewer 2: ```suggestion
   The scheduler can only check for lateness once a task has appeared in its
   active task window. In Cylc 8 this is usually when the task is actually
   ready to run, which severely limits the usefulness of late events as
   currently implemented.
```

Author: Sphinx doesn't mind, right? Is 3 spaces a convention we've agreed on?

Author: (applied anyway)

Reviewer 2: @oliver-sanders Is it worth adding that to CONTRIBUTING.md?

Reviewer: Yeah, in a slightly more refined form.

I think this is in the Sphinx docs and in the rose/developing page (I think the rose dev pages are a little out of date).

If a real time (clock-triggered) workflow performs fairly consistently from one
cycle to the next, you may want to be notified when certain tasks are running
late with respect the time they normally trigger in each cycle.

Cylc can generate a *late* event if a task has not triggered by a given offset
from its cycle point in real time. For example, if a task ``forecast`` normally
triggers at 30 minutes after cycle point, a late event could be configured like this:

.. code-block:: cylc

   [runtime]
      [[forecast]]
           script = run-model.sh
           [[[events]]]
               late offset = PT40M  # allow a 10 minute delay
               late handler = my-handler %(message)s

.. warning::

>   Late offset intervals are not computed automatically so be careful to update
>   them after any workflow change that affects triggering times.

Reviewer 2: ```suggestion
   Late offset intervals are not computed automatically so be careful to update
   them after any workflow change that affects triggering times.
```

Workflow Events
^^^^^^^^^^^^^^^

Cylc also supports general and event-specific workflow event handling so
you can run an event handler or send and email when the scheduler starts up or
shuts down, for example.

> Workflow events are configured in :cylc:conf:`workflow events
> <[scheduler][events]>`. The list of events is:

Reviewer 2: ```suggestion
Workflow events are configured in :cylc:conf:`[scheduler][events]`.
The list of events is:
```

Author: oops :+1:
